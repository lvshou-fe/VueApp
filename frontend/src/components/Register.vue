<template>
  <div class="register-container">
    <form autocomplete="off" name="register-form" @submit.prevent="onSubmit">
      <fieldset>
        <legend>Account details</legend>
        <div class="form-field">
          <input v-model="model.username" v-validate="'required'" name="username" type="text" placeholder="Username"/>
          <span class="field-error"><span v-show="errors.has('username')">Username is required</span></span>
        </div>
        <div class="form-field">
          <input v-model="model.password" v-validate="'required'" name="password" type="password" placeholder="Password"/>
          <span class="field-error"><span v-show="errors.has('password')">Password is required</span></span>
        </div>
      </fieldset>
      <fieldset>
        <legend>Personal details</legend>
        <div class="form-field">
          <input v-model="model.name.firstName" v-validate="'required'" name="firstName" type="text" placeholder="Firstname"/>
          <span class="field-error"><span v-show="errors.has('firstName')">Firstname is required</span></span>
        </div>
        <div class="form-field">
          <input v-model="model.name.lastName" v-validate="'required'" name="lastName" type="text" placeholder="Lastname"/>
          <span class="field-error"><span v-show="errors.has('lastName')">Lastname is required</span></span>
        </div>
      </fieldset>
      <div class="submit-error">
        <span v-show="submitError" class="form-error">Registration failed, try again.</span>
      </div>
      <button :disabled="disableSubmit" type="submit">Register</button>
    </form>
    <div class="registration-success">
      <span v-if="registrationSuccess">Registration was successful. You can now <router-link to="login">login</router-link>.</span>
    </div>
  </div>
</template>

<script>
export default {
  name: 'register',
  data() {
    return {
      model: {
        username: '',
        password: '',
        name: {
          firstName: '',
          lastName: ''
        }
      },
      submitError: false,
      registrationSuccess: false
    };
  },
  computed: {
    disableSubmit() {
      return !Object.keys(this.fields).some(key => this.fields[key].dirty);
    }
  },
  methods: {
    onSubmit() {
      const handleValidation = (isValid) => {
        return isValid ? this.$store.dispatch('REGISTER', this.model) : null;
      };

      const handleResponse = (response) => {
        if (response) {
          this.registrationSuccess = true;
        } else {
          this.submitError = true;
        }
      };

      this.$validator
        .validateAll()
        .then(handleValidation)
        .then(handleResponse)
        .catch((err) => {
          // eslint-disable-next-line
          console.log('Error. Form validation failed!', err);
        });
    }
  }
};
</script>

<style lang="scss" scoped>
.register-container {
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: center;
  justify-content: center;
}

form[name='register-form'] {
  display: flex;
  flex-direction: column;
  width: 300px;
  z-index: 2;
}

.registration-success {
  height: 22px;
  margin-top: 35px;
}
</style>
