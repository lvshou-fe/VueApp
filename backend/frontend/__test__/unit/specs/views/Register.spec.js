import { createLocalVue, shallow } from '@vue/test-utils';
import VeeValidate from 'vee-validate';
import Register from '@/views/Register';

const localVue = createLocalVue();
localVue.use(VeeValidate);

describe('Register.vue', () => {
  it('should render', () => {
    const wrapper = shallow(Register, {
      localVue
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
