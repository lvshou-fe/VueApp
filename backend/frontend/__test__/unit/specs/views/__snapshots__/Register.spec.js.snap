// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register.vue should render 1`] = `
<div class="register-container">
  <form autocomplete="off" name="register-form">
    <fieldset>
      <legend>Account details</legend>
      <div class="form-field">
        <input name="username" type="text" placeholder="Username" data-vv-id="1" aria-required="true" aria-invalid="false"> <span class="field-error"><span style="display: none;">Username is required</span></span>
      </div>
      <div class="form-field">
        <input name="password" type="password" placeholder="Password" data-vv-id="2" aria-required="true" aria-invalid="false"> <span class="field-error"><span style="display: none;">Password is required</span></span>
      </div>
    </fieldset>
    <fieldset>
      <legend>Personal details</legend>
      <div class="form-field">
        <input name="firstName" type="text" placeholder="Firstname" data-vv-id="3" aria-required="true" aria-invalid="false"> <span class="field-error"><span style="display: none;">Firstname is required</span></span>
      </div>
      <div class="form-field">
        <input name="lastName" type="text" placeholder="Lastname" data-vv-id="4" aria-required="true" aria-invalid="false"> <span class="field-error"><span style="display: none;">Lastname is required</span></span>
      </div>
    </fieldset>
    <div class="submit-error"><span class="form-error" style="display: none;">Registration failed, try again.</span></div>
    <button disabled="disabled" type="submit">Register</button>
  </form>
  <div class="registration-success">
    <!---->
  </div>
</div>
`;
